docs/newtypes/index.md

```diff
@@ -0,0 +1,310 @@
+
+In many cases that makes sense. We often want to use a new type to denote that this particular `Int` value represents a sequence number and shouldn't be confused with another `Int` value that represents an account number without imposing other constraints.
+
+However, in other cases we may want to restrict the values that the underlying type can take. For instance, perhaps a `SequenceNumber` should never be negative.
+
+We can model this in ZIO Prelude by extending `NewtypeSmart` and `SubtypeSmart` when creating our new type and providing an `Assertion` that describes the constraints on the underlying value. We call these types smart types because they build in the logic for their own validation.
+
+```scala mdoc:reset-object
+import zio.prelude.SubtypeSmart
+import zio.test.Assertion._
+
+object SequenceNumber extends SubtypeSmart[Int](isGreaterThanEqualTo(0))
+type SequenceNumber = SequenceNumber.Type
+```
+
+When we create a smart type the `apply`, `wrap`, and `wrapAll` operators will be `protected` and will only be accessible within the scope of the smart type object unless we choose to expose them. This prevents the users of our smart type from creating invalid instances of the smart type.
+
+```scala mdoc:reset-object
+import zio.prelude.{SubtypeSmart, Validation}
+import zio.test.Assertion._
+
+object SequenceNumber extends SubtypeSmart[Int](isGreaterThanEqualTo(0)) {
+  val initial: SequenceNumber =
+    SequenceNumber(0) // okay
+}
+type SequenceNumber = SequenceNumber.Type
+
+// val invalid: SequenceNumber =
+//   SequenceNumber(-1) // does not compile
+```
+
+Instead, the `make` or `makeAll` operators can be used to construct instances of the smart type with validation.
+
+```scala mdoc
+import zio.Chunk
+
+val validatedSequenceNumber: Validation[String, SequenceNumber] =
+  SequenceNumber.make(1)
+
+val validateSequenceNumbers: Validation[String, Chunk[SequenceNumber]] =
+  SequenceNumber.makeAll(Chunk(1, 2, 3))
+```
+
+We can see that `make` and `makeAll` return a `Validation` that will either be a success with a valid instance of the smart type or a failure with a helpful error message indicating why a valid instance of the smart type could not be constructed.
+
+Because we have access to the `apply`, `wrap`, and `wrapAll` constructors within the smart type object we can implement operators that allow us to create instances of the smart type internally with zero overhead when we know that the underlying data is valid.
+
+For example, we might want to expose an `unsafeMake` operator for our users to allow them to construct instances of the smart type directly without going through `Validation`. As its name implies this places responsibility on the user to ensure that the underlying data is valid but could be necessary for performance or ergonomics in some situations.
```

Comment:
I kind of like this as well. I was constructing ad hoc values in some tests and was annoyed to unwrap a `Validation` in the test with not much of a reason to do so.

Reply:
You may give an example with custom assertions or bring up why they may be useful.

Reply:
Good idea. Added a couple sentences about that.